src/coreclr/jit/llvm.cpp

```diff
@@ -1587,16 +1607,28 @@ void Llvm::buildReturn(GenTree* node)
 
 void Llvm::buildStoreInd(GenTreeStoreInd* storeIndOp)
 {
-    Value* address = getGenTreeValue(storeIndOp->Addr());
-    Value* toStore = getGenTreeValue(storeIndOp->Data());
-    if (toStore->getType()->isPointerTy() && (storeIndOp->gtFlags & GTF_IND_TGT_NOT_HEAP) == 0)
-    {
-        // RhpAssignRef will never reverse PInvoke, so do not need to store the shadow stack here
-        _builder.CreateCall(getOrCreateRhpAssignRef(), ArrayRef<Value*>{address, castIfNecessary(toStore, Type::getInt8PtrTy(_llvmContext))});
-    }
-    else
+    GenTree* data = storeIndOp->Data();
+    Value* toStore = getGenTreeValue(data);
+    Value* address = consumeValue(storeIndOp->Addr(), toStore->getType()->getPointerTo());
+
+    GCInfo gcInfo(_compiler);
+    GCInfo::WriteBarrierForm writeBarrierForm = gcInfo.gcIsWriteBarrierCandidate(storeIndOp, data);
+    switch (writeBarrierForm)
     {
-        castingStore(toStore, address, storeIndOp->gtType);
+        case GCInfo::WriteBarrierForm::WBF_BarrierUnchecked:
+            _builder.CreateCall(getOrCreateRhpAssignRef(),
+                                ArrayRef<Value*>{castIfNecessary(address, Type::getInt8PtrTy(_llvmContext)),
+                                                 castIfNecessary(toStore, Type::getInt8PtrTy(_llvmContext))});
+            break;
+        case GCInfo::WriteBarrierForm::WBF_BarrierChecked:
+        case GCInfo::WriteBarrierForm::WBF_BarrierUnknown:
+            _builder.CreateCall(getOrCreateRhpCheckedAssignRef(),
+                                ArrayRef<Value*>{castIfNecessary(address, Type::getInt8PtrTy(_llvmContext)),
+                                                 castIfNecessary(toStore, Type::getInt8PtrTy(_llvmContext))});
+            break;
+        default:
+            castingStore(toStore, address, storeIndOp->gtType);
```

Comment:
It would be better to have an explicit `WBF_NoBarrier` case for this and then `unreached` for the `default`.

Also, we should `consumeValue(data, getLlvmTypeForVarType(storeIndOp->TypeGet()))` for `toStore` and delete `castingStore` (you can also delete `storeOnShadowStack` that is its only other user, and which itself is unused).